dex/elrond_dex_pair/src/sharer.rs

```diff
@@ -0,0 +1,326 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use crate::liquidity_pool;
+use elrond_wasm::elrond_codec::TopEncode;
+use multitransfer::EsdtTokenPayment;
+
+use super::amm;
+use super::config;
+
+const BP: u64 = 100_000;
+const GAS_COST_FOR_SEND_LIQUIDITY: u64 = 100_000_000u64;
+
+#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, PartialEq, Clone)]
+pub struct SharedInformation<BigUint: BigUintApi> {
+    pub metadata: InformationMetadata,
+    pub liquidity_info: LiquidityInformation<BigUint>,
+    pub swap_stats: SwapStatistics<BigUint>,
+}
+
+#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, PartialEq, Clone)]
+pub struct InformationMetadata {
+    pub sender: Address,
+    pub timestamp: u64,
+}
+
+#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, PartialEq, Clone)]
+pub struct LiquidityInformation<BigUint: BigUintApi> {
+    pub liquidity_amount: BigUint,
+    pub first_token_amount: BigUint,
+    pub second_token_amount: BigUint,
+}
+
+#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, PartialEq, Clone)]
+pub struct SwapStatistics<BigUint: BigUintApi> {
+    pub _placeholder: BigUint,
+}
+
+impl<BigUint: BigUintApi> SharedInformation<BigUint> {
+    pub fn to_boxed_bytes(&self) -> BoxedBytes {
+        let mut vec = Vec::new();
+        let result = self.top_encode(&mut vec);
+        match result {
+            Result::Ok(_) => BoxedBytes::from(vec.as_slice()),
+            Result::Err(_) => BoxedBytes::empty(),
+        }
+    }
+
+    pub fn from_boxed_bytes(bytes: BoxedBytes) -> SCResult<SharedInformation<BigUint>> {
+        SharedInformation::<BigUint>::top_decode(bytes.as_slice()).into()
+    }
+}
+
+#[elrond_wasm::module]
+pub trait SharerModule:
+    info_sync::InfoSyncModule
+    + amm::AmmModule
+    + config::ConfigModule
+    + token_supply::TokenSupplyModule
+    + token_send::TokenSendModule
+    + liquidity_pool::LiquidityPoolModule
+    + multitransfer::MultiTransferModule
+{
+    #[endpoint(shareInformation)]
+    fn share_information(&self) -> SCResult<()> {
+        let block = self.blockchain().get_block_nonce();
+
+        if block > self.last_info_share_block().get() + self.info_share_min_blocks().get() {
```

Comment:
we must share information or let information share if the difference between local and virtual view is too much